### PR TITLE
Add test to fire scrollend events for user scrolls

### PR DESCRIPTION
**Pull request by cathiechen**
Hi,
This patch adds scrollend event tests to user scrolls: dragging scroll, pressing scrollbar scroll, dragging scrollbar thumb scroll, and sending DOWN key scroll.

### CODE REVIEW

File: dom/events/scrolling/scroll_support.js [hunk 1]
```diff
@@ -83,3 +83,18 @@ function touchScrollInTarget(pixels_to_scroll, target, direction, pause_time_in_
...
+  return new test_driver.Actions()
+    .addPointer("pointer1", "mouse")
+    .pointerMove(origin.x, origin.y, { origin: target })
+    .pointerDown()
+    .pointerMove(origin.x + delta.x, origin.y + delta.y, { origin: target })
+    .pointerMove(origin.x + delta.x * 2, origin.y + delta.y * 2, { origin: target })
+    .send()
+    .then(() => {
+      setTimeout(() => {
+        return new test_driver.Actions().pointerUp().send();
```
Comment (NavidZ) [inline dom/events/scrolling/scroll_support.js:97]: Instead of this why not just use .pause function on the Actions object? Like
`.pointermove(...).pause(pause_time_in_ms).pointerUp().send();`

But that aside why did you need this delay in the first place?

File: dom/events/scrolling/scrollend-event-for-user-scroll.html [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+function runTest() {
+    promise_test(async (t) => {
+        // Make sure that no scrollend event is sent to window.
+        window.addEventListener("scrollend",
+            t.unreached_func("window got unexpected scrollend event."));
+        await waitForCompositorCommit();
+        scrollend_arrived = false;
+
+        // Perform drag actiion on target div and wait for target_div to get scrollend event.
```
Comment (NavidZ) [inline dom/events/scrolling/scrollend-event-for-user-scroll.html:65]: actiion -> action

File: dom/events/scrolling/scrollend-event-for-user-scroll.html [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        var origin = { x: target_div.offsetWidth / 2, y: target_div.offsetHeight - 50 };
+        var delta = { x: 0, y: 40 };
+        await mouseActionsInTarget(target_div, origin, delta, 300);
+        assert_false(scrollend_arrived);
```
Comment (NavidZ) [inline dom/events/scrolling/scrollend-event-for-user-scroll.html:69]: Why do you expect this? I mean if the scrolling is done this might even be true after the mouseActionsInTarget promise is resolved.

File: dom/events/scrolling/scrollend-event-for-user-scroll.html [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        // Make sure that no scrollend event is sent to window.
+        window.addEventListener("scrollend",
+            t.unreached_func("window got unexpected scrollend event."));
+        await waitForCompositorCommit();
+        scrollend_arrived = false;
+
+        // Perform drag actiion on target div and wait for target_div to get scrollend event.
+        var origin = { x: target_div.offsetWidth / 2, y: target_div.offsetHeight - 50 };
+        var delta = { x: 0, y: 40 };
+        await mouseActionsInTarget(target_div, origin, delta, 300);
```
Comment (NavidZ) [inline dom/events/scrolling/scrollend-event-for-user-scroll.html:68]: qq. Why this mouse drag on div causes scrolling? Is it happening on the scrollbar or something?

File: dom/events/scrolling/scrollend-event-for-user-scroll.html [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    promise_test(async (t) => {
+        resetTargetScrollState();
+        // Make sure that no scrollend event is sent to window.
+        window.addEventListener("scrollend",
+            t.unreached_func("window got unexpected scrollend event."));
+        await waitForCompositorCommit();
+        scrollend_arrived = false;
+
+        // Drag the thumb of target div.
+        var origin = { x: target_div.offsetWidth - 5, y: 50 };
```
Comment (NavidZ) [inline dom/events/scrolling/scrollend-event-for-user-scroll.html:103]: I assume the scrollbars are not fully specified. So they might be overlay or not have this much width. I'm not sure what should be done in that case. @bokand do you know what can be done with respect to this in testing?

File: dom/events/scrolling/scroll_support.js (hunk 1 above)
Comment (cathiechen) [reply dom/events/scrolling/scroll_support.js:97]: Sorry, I was working on WebKit. It seems the test_driver interfaces in WebKit does not update. 
In the old interfaces, `dispatchMouseActions()` in resources/testdriver-vendor.js omits `pause`.
```
function dispatchMouseActions(actions)
{
    if (!window.eventSender)
        return Promise.reject(new Error("window.eventSender is undefined."));

    return new Promise(resolve => {
        setTimeout(() => {
            eventSender.dragMode = false;

            for (let action of actions) {
                switch (action.type) {
                case "pointerMove":
                    let origin = { x: 0, y: 0 };
                    if (action.origin instanceof Element) {
                        const bounds = action.origin.getBoundingClientRect();
                        logDebug(() => `${action.origin.id} [${bounds.left}, ${bounds.top}, ${bounds.width}, ${bounds.height}]`);
                        origin.x = bounds.left + 1;
                        origin.y = bounds.top + 1;
                    }
                    logDebug(() => `eventSender.mouseMoveTo(${action.x + origin.x}, ${action.y + origin.y})`);
                    eventSender.mouseMoveTo(action.x + origin.x, action.y + origin.y);
                    break;
                case "pointerDown":
                    logDebug(() => `eventSender.mouseDown()`);
                    eventSender.mouseDown(action.button);
                    break;
                case "pointerUp":
                    logDebug(() => `eventSender.mouseUp()`);
                    eventSender.mouseUp(action.button);
                    break;
                case "pause":
                    // FIXME: What should we do here?
                    break;
                default:
                    return Promise.reject(new Error(`Unknown action type "${action.type}".`));
                }
            }
            resolve();
        });
    });
}
```
Without `pause`, we can't generate effective drag actions, so I used `setTimeout`here.

And I tried to confirm `pause` is properly handled in the latest code, but I couldn't find it. Sorry...
But I agree to change it to `pause(pause_time_in_ms).pointerUp().send();` here.

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 2 above)
Comment (cathiechen) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:65]: Done!

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 4 above)
Comment (cathiechen) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:68]: This is to simulate the dragging selection, like select all the text content inside this div by dragging the mouse.

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 3 above)
Comment (cathiechen) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:69]: OK, make sense. I'll remove this.

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 5 above)
Comment (cathiechen) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:103]: Maybe we can use offsetWidth - clientWidth to calculate the scrollbar width?

**Comment by cathiechen** [comment]
Hi @NavidZ ,
I think this patch is ready for review. Please take a look. Thanks:)

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 4 above)
Comment (NavidZ) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:68]: Interesting scenario. I didn't think of it when writing the spec.  Do we send scroll events in these cases. Either way, I'm fine with this and it sounds reasonable.

**Review by NavidZ — APPROVED** [review]

File: dom/events/scrolling/scrollend-event-for-user-scroll.html (hunk 4 above)
Comment (cathiechen) [reply dom/events/scrolling/scrollend-event-for-user-scroll.html:68]: Yes, there are scroll events when dragging.
BTW, I guess one common scenario is selecting the content and copy:)
